.github/workflows/test.yml

```diff
@@ -0,0 +1,156 @@
+        uses: actions/setup-go@v3
+        with:
+          go-version: "${{ steps.get-go-version.outputs.go-version }}"
+      - name: Determine go cache key
+        id: go-cache-key
+        run: |
+          echo "key=${{ runner.os }}-go-${{ hashFiles('**/go.sum', './Makefile', './tools/tools.go') }}" >> $GITHUB_OUTPUT
+      - name: Determine Go cache paths
+        id: go-cache-paths
+        run: |
+          echo "go-build=$(go env GOCACHE)" >> $GITHUB_OUTPUT
+          echo "go-mod=$(go env GOMODCACHE)" >> $GITHUB_OUTPUT
+          echo "go-bin=$(go env GOPATH)/bin" >> $GITHUB_OUTPUT
+      - name: Set up Go modules cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            ${{ steps.go-cache-paths.outputs.go-build }}
+            ${{ steps.go-cache-paths.outputs.go-mod }}
+            ${{ steps.go-cache-paths.outputs.go-bin }}
+          key: ${{ steps.go-cache-key.outputs.key }}
+          restore-keys: |
+            ${{ runner.os }}-go
+      - name: Install Tools
+        run: |
+          go mod download
+          make tools
+  test-modules:
+    needs:
+      - setup
+    runs-on: ${{ fromJSON(vars.RUNNER) }}
+    strategy:
+      matrix:
+        module: ["api", "sdk"]
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up go
+        uses: actions/setup-go@v3
+        with:
+          go-version: "${{ needs.setup.outputs.go-version }}"
+      - name: Set up Go modules cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            ${{ needs.setup.outputs.cache-go-build }}
+            ${{ needs.setup.outputs.cache-go-mod }}
+            ${{ needs.setup.outputs.cache-go-bin }}
+          key: ${{ needs.setup.outputs.go-cache-key }}
+          restore-keys: |
+            ${{ runner.os }}-go
+      - name: Test ${{ matrix.module }} Module
+        run: |
+          make test-${{ matrix.module }}
+
+  set-test-package-matrix:
+    runs-on: ${{ fromJSON(vars.RUNNER) }}
+    needs:
+      - setup
+    outputs:
+      package-matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up go
+        uses: actions/setup-go@v3
+        with:
+          go-version: "${{ needs.setup.outputs.go-version }}"
+      - name: Set up Go modules cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            ${{ needs.setup.outputs.cache-go-build }}
+            ${{ needs.setup.outputs.cache-go-mod }}
+            ${{ needs.setup.outputs.cache-go-bin }}
+          key: ${{ needs.setup.outputs.go-cache-key }}
+          restore-keys: |
+            ${{ runner.os }}-go
+      - id: set-matrix
+        run: ./.github/scripts/set-test-package-matrix.sh
+
+  test:
+    needs:
+      - setup
+      - set-test-package-matrix
+    runs-on: ${{ fromJSON(vars.RUNNER) }}
+    strategy:
+      fail-fast: false
+      matrix:
+        package: ${{ fromJson(needs.set-test-package-matrix.outputs.package-matrix) }}
```

Comment:
Very nice. Does this run one job per package? If so, that sounds like we'll do a lot of setup for what could be a small amount of tests?

Reply:
Looks like this test run spent 4 minutes building boundary and 0.1s running the tests: https://github.com/hashicorp/boundary/actions/runs/4245687164/jobs/7382157690.

Reply:
Yea, specifically that is building the plugins, not all of boundary. I think we could move that plugin build step up prior to the matrix and cache it like we do the go mod cache. I would note that we have several packages that take >10m to run though. So as long as jobs are not waiting for runners, we are always limited by the slowest package.

Reply:
I'm generally ecstatic about this new run strategy, I think it'll fix a lot of issues around test timeouts and CI execution times, don't get me wrong! I wonder if it'd be worth amending the script in a way that we can group together known fast-running tests in a way that any new packages automatically get their own matrix coordinate. That way, we can amortize the cost of things like the postgres startup for tests that we know run fast.

Reply:
I think those can be follow up tasks if need be. The primary goal here is to get this running as a github action, we can do additional optimizations later.

Reply:
Totally fair 😄